build: accept -f - to read Dockerfile from stdin

tonistiigi opened the pull request:
fixes #27393

Heavily based on implementation by @dsheets https://github.com/dsheets/docker/commits/build-dockerfile-stdin

Signed-off-by: David Sheets <sheets@alum.mit.edu>
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

dsheets committed:
build: accept -f - to read Dockerfile from stdin

Heavily based on implementation by David Sheets

Signed-off-by: David Sheets <sheets@alum.mit.edu>
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

aaronlehmann commented on pkg/archive/archive.go:
```diff
@@ -225,6 +225,60 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
 }
 
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// with a single file called header.Name with contents of file.
```
The comment is a bit unclear. It makes it sound like the contents of the existing tar is replaced with just one file.

aaronlehmann commented on pkg/archive/archive.go:
```diff
@@ -225,6 +225,60 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
 
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// with a single file called header.Name with contents of file.
+// TODO: make this into a generic tar conversion function with walkFn argument
```
Agreed, this seems very similar to `replaceDockerfileTarWrapper`.

aaronlehmann replied:
Also, I would rather avoid having two nested tar stream modifications in the trusted + stdin case.

tonistiigi replied:
Yes, it could be like a pipeline of modifications. The same pipeline steps could be also used for tar creation and extraction. Atm there is big list of specific arguments, for example user remapping, filename renaming for cp etc. while all these could be implemented as a conversion step.

aaronlehmann commented:
Design LGTM

yallop commented:
Why are the commits squashed here?  It'd be better to have two commits, to preserve history and authorship.  Could you rebase to split the commits, @tonistiigi?

dmcgowan commented on pkg/archive/archive.go:
```diff
@@ -225,6 +225,61 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+				}
+
+				tarWriter.Close()
+				pipeWriter.Close()
```
Not sure how important it is for how this tar stream is used but after reading the tar content there could be remaining data that on the input tar stream, we often pipe that to null to ensure that anything blocking on that read is freed (without needing to rely on the close).

tonistiigi replied:
You mean in case of an error? Can `tarReader` really skip data if it returns `io.EOF`

dmcgowan commented on pkg/archive/archive.go:
```diff
@@ -225,6 +225,61 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+			if err == io.EOF {
+				header.Size = int64(len(file))
+
+				if err := tarWriter.WriteHeader(&header); err != nil {
```
Instead of always writing at end, write when `err == io.EOF` or as soon as `header.Name <= hdr.Name`

tonistiigi replied:
I've updated this and added more tests.

thaJeztah commented:
We discussed this PR in the maintainers meeting and like this feature; @vieux is gonna look into adding the same for `-f /path/to/a/Dockerfile` as a follow up

dmcgowan commented on pkg/archive/archive.go:
```diff
@@ -225,6 +225,73 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+				return
+			}
+
+			content := io.Reader(tarReader)
```
What is the point of this assignment instead of just directly passing into io.Copy or assigning here. I see this as unnecessarily shadowing the arg `content`.

duglin commented on cli/command/image/build.go:
```diff
@@ -235,6 +247,28 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		}
 	}
 
+	// replace Dockerfile if added dynamically
```
If someone already has a Dockerfile  but its not being used for the build, and they use "-" on the `-f`, won't this replace their Dockerfile and cause incorrect results if their Dockerfile is used for some other purpose?

duglin replied:
I think this comment needs to be updated based on the recent change you made to generate a random dockerfile name.

duglin commented:
While I like the idea of this, I'm not comfortable with replacing a user-owned file with the stdin and we should be adding any files we inject into the context to .dockerignore.

My preferred solution would be to make the input of the API a multi-part mime where one section is the tar and another section is the Dockerfile.  But, that's a much bigger change.  Short of that, can we consider storing the stdin into a new/random file that then gets added to .dockerignore?  Our goal should be to have zero impact on existing files in the context.  Meaning, a Dockerfile cmd of `COPY * ...` should not see any new files due to this feature.

tonistiigi commented:
@duglin This is definitely harder but I'll see what I can do. Although for `-` I don't see much of a problem. `Dockerfile` is the default name that is always used unless user sets a new name with `-f`. If they set `-f -` they don't really specify the new name. But I can see it becoming more complicated when we start accepting absolute paths outside working dir for `-f` as well.

cpuguy83 commented:
Can't we write to some temp file within the context, add to dockerignore, and set the dockerfile path to the temp file when sending to the API? If the user wants it to be part of their context they shouldn't use stdin, I think?

tonistiigi commented:
@cpuguy83 I guess that's what @duglin was suggesting as well. Note though that even atm we don't make any promises about the contents of `Dockerfile`. Using content trust with `COPY Dockerfile .` you would not get the same contents.

duglin commented:
@tonistiigi can you elaborate on your statement about the Dockerfile?  With `docker build -f foo .` I would hope that any file in `.` would be totally untouched, including one named `Dockerfile`. So the user could put anything they wanted in there an be assured that docker won't touch it.

duglin commented:
To be honest, I'm not thrilled with modifying the user's .dockerignore file either since we never claimed it was "ours". To me its the user's input to us.

Another option would be to add a flag to **just** the API for now (e.g. `&tmpDockerfile`) to indicate that the referenced Dockerfile is a temporary one that we created that should be deleted after it is read.  Then we know for sure we're not mucking with the user's files at all.

tonistiigi commented:
@duglin When content trust is used we add a digest to every `FROM` instruction if needed, before uploading the Dockerfile. We don't rename the file while doing so.

duglin commented:
that's interesting - is this during a `docker build` so that it know which file is the real Dockerfile?

tonistiigi commented:
@duglin That's in https://github.com/docker/docker/blob/3fe2730ab3d7bba483d7bf4fb1e95b627823ff4d/cli/command/image/build.go#L384-L388

tonistiigi commented:
@duglin @cpuguy83 I have updated to use the random name and `.dockerignore`. I generalized the tar converter method as well so the same method can be more easily used for dct and `plugin create`.

duglin commented on cli/command/image/build.go:
```diff
@@ -235,6 +247,47 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
+			AccessTime: now,
+			ChangeTime: now,
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:12]
```
I know its really really unlikely, but to avoid a flaky situation, can you just check to make sure there isn't a file with this name already in the build context? And if so, try again.

tonistiigi replied:
In 99% of the places we don't check things like this and it would make the code really complicated with multiple passes to achieve this. I increased the size if you are actually worried about the collisions.

duglin commented on cli/command/image/build/context.go:
```diff
@@ -219,6 +224,8 @@ func getDockerfileRelPath(givenContextDir, givenDockerfile string) (absContextDi
 				absDockerfile = altPath
 			}
 		}
+	} else if absDockerfile == "-" {
```
would this ever be true? I'm wondering if we would see the random name instead. And if so, then I think we can remove the "if" on line 243 below (`if givenDockerfile != "-" {`) since it would always be true.

tonistiigi replied:
This is called before the name replacement. So the return value is replaced anyway, but would be nice not to return anything broken from the func. So it returns the default as it can't determine the relative Dockerfile path.

duglin commented on cli/command/image/build.go:
```diff
@@ -235,6 +247,47 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
+						return nil, nil, err
+					}
+				}
+				b.Write([]byte("\n" + randomName))
```
first a nit... can you add `\n` to the end of this line so that if someone does a `cat .dockerignore` it doesn't look like its missing a CR at the end.

More importantly, iff .dockerignore did not already exist then I think we need to also add `.dockerignore` to `.dockerignore`.  This is so that we're not leaving around extra files in the user's context that they don't want - even hidden files.  Granted we should also try to remove these extra lines from the file that we're adding but that's a lot harder and at this time probably not worth the effort - until someone complains. 

Also, in my testing I'm seeing the random file appear but only when I already have a `.dockerignore` with `.dockerignore` in it:
```
$ docker build -f - . < ddd
Sending build context to Docker daemon  6.144kB
Step 1/3 : FROM ubuntu
 ---> 0ef2e08ed3fa
Step 2/3 : COPY * /tmp/
 ---> 93546ad88407
Removing intermediate container 9a87cad3d2bd
Step 3/3 : RUN ls -alt /tmp/
 ---> Running in 278f61144204
total 20
drwxr-xr-x 35 root root 4096 Mar 10 13:34 ..
drwxrwxrwt  2 root root 4096 Mar 10 13:34 .
-rw-------  1 root root   46 Mar 10 13:34 .dockerfile.3146c886e77c
-rw-r--r--  1 root root   46 Mar 10 13:34 ddd
-rw-r--r--  1 root root   19 Mar 10 13:25 Dockerfile
 ---> 2b8485d37faf
Removing intermediate container 278f61144204
Successfully built 2b8485d37faf

$ cat ddd
FROM ubuntu
COPY * /tmp/
RUN ls -alt /tmp/
```

Interestingly, when I add Dockerfile to .dockerignore its not removed either - and this does appear to work on master so something in this PR changed the .dockerignore processing logic.

tonistiigi replied:
Fixed.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -2022,6 +2022,53 @@ func (s *DockerSuite) TestBuildNoContext(c *check.C) {
+	}).Assert(c, icmd.Success)
+
+	out, _ := dockerCmd(c, "run", "--rm", name, "ls", "/baz")
+	c.Assert(strings.TrimSpace(string(out)), checker.Equals, "foo")
```
I like this test, but once the other issues are fixed can we extend this to make sure:
1 - that .dockerignore is removed too when its not already there
2 - that .dockerignore is kept around if it was present before the build cmd

duglin commented on pkg/archive/archive.go:
```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+// while replacing a single file called header.Name with new contents.
+// If the file with header.Name does not exist it is added to the tar stream.
+// TODO: make this into a generic tar conversion function with walkFn argument
+func ReplaceFileTarWrapper(inputTarStream io.ReadCloser, mods map[string]TarModifierFunc) io.ReadCloser {
```
@vbatts would you have time to look this part of the PR over?

vbatts replied:
largely looks alright. I haven't tested it though.
Some of the gotchas are to ensure that the caller _always_ closes the returned `io.ReadCloser` so there aren't goroutines floating around. Might be good to call that out in the comment.

duglin commented:
I know its a major edge case but this PR does change current behavior if someone used "-" as their Dockerfile name. And I'm ok with the change for the greater good.  However, since "-" as a Dockerfile name is now a "special case" for the CLI I'm thinking that the daemon/API side of things should reject a build API call with "-" as the Dockerfile name. Not rejecting it would mean that there's an inconsistency between valid Dockerfile names when using the CLI vs the API and that's probably not good.  If needed, both paths could get around this by using `./-` and we'd still be consistent between the two w.r.t. what's an acceptable name.   Thoughts?

duglin commented:
Another thing... not necessarily for this PR though... you mentioned that DCT will muck with the Dockerfile's FROM, once this PR is merged we should modify that DCT code to create a new random Dockerfile just like you're doing here so that the user is totally unaware of us changing their Dockerfile on them.  Unless we really do want the user to see the sha of the image we think should be used.  hmmm now I'm not sure.....  just something to ponder

tonistiigi commented:
I added the tests for other `.dockerignore` cases and fixed an issue in current master where Dockerfile and `.dockerignore` would be out of order in the sent context. Another weird thing I saw is that daemon only removes dockerfile and dockerignore if they are in `.dockerignore` and doesn't touch any other files. But I didn't want to change that atm. PTAL.

> I'm thinking that the daemon/API side of things should reject a build API call with "-" as the Dockerfile name.

Why would API handle `-` any differently as there is no concept of stdin for API? It is a client side replacement. The API should work with any CLI.

dmcgowan commented on pkg/archive/archive.go:
```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+				return
+			}
+
+			if _, err := io.Copy(tarWriter, tarReader); err != nil {
```
Doesn't look like `tar.Reader` or `tar.Writer` implement `io.ReaderFrom` or `io.WriterTo`, might be better to use `io.CopyBuffer` to avoid making a 32K buffer for each call.

cpuguy83 replied:
or `pools.Copy`

tonistiigi replied:
There is only 1 iteration of this in the regular case. Maximum 2 for the weird `.dockerignore` case.

tonistiigi replied:
Thanks. Updated.

thaJeztah commented:
ping @duglin @aaronlehmann what's the status on this one?

re: https://github.com/docker/docker/pull/31236#issuecomment-281656499

> Why are the commits squashed here? It'd be better to have two commits, to preserve history and authorship. Could you rebase to split the commits,

@tonistiigi could you split the commits if possible?

aaronlehmann commented on cli/command/image/build.go:
```diff
@@ -227,13 +239,56 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		buildCtx, err = archive.TarWithOptions(contextDir, &archive.TarOptions{
 			Compression:     compression,
 			ExcludePatterns: excludes,
-			IncludeFiles:    includes,
```
Why the change from `IncludeFiles` to negative exclusions?

tonistiigi replied:
Includes array mixes up the order of files. Extra includes get new walkers after after the actual context has been parsed.

aaronlehmann replied:
Please mention this in a comment.

dnephin replied:
This is the only issue I wasn't sure about. What do you think the comment should say?

aaronlehmann replied:
I just wanted it to explain why `ExcludePatterns` is being used instead of `IncludeFiles` for additional files that are added to the tar. According to @tonistiigi above, it's because `IncludeFiles` messes up the order of the files. That said, I don't know if we care about file ordering anymore.

tonistiigi replied:
If they are equal we should still keep the order. This applies even if `Dockerfile` does not come from stdin.

dnephin replied:
Ok, added another line to this comment about ordering.

aaronlehmann commented on cli/command/image/build.go:
```diff
@@ -227,13 +239,56 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:20]
+
+		buildCtx = archive.ReplaceFileTarWrapper(buildCtx, map[string]archive.TarModifierFunc{
```
Now that `archive.ReplaceFileTarWrapper` is more generic, would it be difficult to remove `replaceDockerfileTarWrapper` and just have a single call to `archive.ReplaceFileTarWrapper` when necessary (either trusted build, Dockerfile provided on stdin, or both)? It would be nice to simplify things by removing `replaceDockerfileTarWrapper`, which seems like basically duplicate code at this point, and it would be more efficient and easier to understand if we avoid mulitple layers of wrapping. This would also address @duglin's request to use a randomized Dockerfile name when doing the trusted-build reference rewriting.

tonistiigi replied:
I agree but that should be in next PR. We have more places than `replaceDockerfileTarWrapper` that use this pattern.

aaronlehmann commented on cli/command/image/build.go:
```diff
@@ -227,13 +239,56 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
+				} else {
+					extraIgnore += ".dockerignore\n"
+				}
+				b.Write([]byte("\n" + extraIgnore))
```
It probably doesn't matter, but should the leading `\n` be written inside `if content != nil`?

dnephin replied:
I've changed the logic here to write to the buffer directly instead of using the `extraIgnore` string. I think the idea here was that you don't know if their content ended with a newline, so you have to add another here to be sure.

aaronlehmann commented on pkg/archive/archive.go:
```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// while replacing a single file called header.Name with new contents.
+// If the file with header.Name does not exist it is added to the tar stream.
+// TODO: make this into a generic tar conversion function with walkFn argument
```
Comment needs to be updated.

aaronlehmann commented on pkg/archive/archive.go:
```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+	loop0:
+		for {
+			hdr, err := tarReader.Next()
+			for len(modKeys) > 0 && (err == io.EOF || err == nil && hdr.Name >= modKeys[0]) {
```
This seems to assume that the input tar stream is lexically sorted, but I don't understand how that assumption is justified.

tonistiigi replied:
https://github.com/docker/docker/pull/31236/files/f9aeef26c280e62f8765c4111c93869d77afedd5#diff-45e15b27a574e8c07552aad3688b3142 Can't have both. We can always mess up the order by adding new files to the end or do the correct thing 99.9% of the cases.

tonistiigi replied:
wrong link https://github.com/docker/docker/pull/31236#pullrequestreview-23351180

aaronlehmann replied:
I think I'm missing some context, but this just seems wrong to me. If the tar input tar stream is unsorted (and there's no particular reason it should be sorted), we'll insert duplicate files that aren't appropriate modifications of the existing files. I would add files either in place of the existing entries, or at the end of the stream if they didn't exist in the input. That seems like the only sane way to do it in a single-pass wrapper.

tonistiigi replied:
cc @dmcgowan 
@aaronlehmann The tarstream we generate is always in order. It is not only in the case when existing tar is used in stdin(in that case we don't support rewriting Dockerfile).

aaronlehmann replied:
It seems like we're depending on an implementation detail of another part of the code, which feels fragile to me. This will also be a problem if/when we need to rewrite things in a user-provided tar, for example rewriting the Dockerfile for content trust in that case. I think it's better to rewrite the files in-place, and insert the missing ones at the end of the stream.

dmcgowan replied:
Discussed this with @tonistiigi some earlier. I agree we should just add the new files at the end since the location in the tar stream should have no impact on the final result and the risk of encountering the file later on out of order would lead to incorrect results.

dnephin replied:
I've updated this function to iterate over all the files and not assume any sorting.

aaronlehmann commented on pkg/archive/archive.go:
```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
+			for len(modKeys) > 0 && (err == io.EOF || err == nil && hdr.Name >= modKeys[0]) {
+				var h *tar.Header
+				var rdr io.Reader
+				if hdr != nil && hdr.Name == modKeys[0] {
```
Can we assume that `hdr` has valid data if it's non-nil? It might be safer to add `err == nil`to this check.

dnephin committed:
Factor out adding dockerfile from stdin.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Upadte archive.ReplaceFileTarWrapper() to not expect a sorted archive

Improve test coverage of ReplaceFileTarWrapper()

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
I've rebased and pushed a couple commits to address these comments.

aaronlehmann commented on cli/command/image/build.go:
```diff
@@ -338,6 +356,50 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
+					return nil, nil, err
+				}
+			} else {
+				b.WriteString("\n.dockerignore")
```
In this case, `content` is nil, so this will be the first entry in the file, and we shouldn't need to begin with a `\n`.

aaronlehmann replied:
...but we should probably include a `\n` at the end in case someone tries to look at the file from the context. I think @duglin requested this in a comment on an earlier version.

dnephin replied:
Outside of this `if/else` there is a write that adds the trailing newline. So I think that's covered.

I'll remove the first `\n`, you're right it's not necessary.

dnephin commented:
Ok, added the comment, and also fixed a failing test. `content` was being passed as an empty reader instead of `nil`. This case is now also caught by the unit tests, and the comment for `TarModifierFunc` mentions that `content` will be `nil` when the file doesn't exist.

aaronlehmann commented:
LGTM

tonistiigi commented:
LGTM for the latest updates

vdemeester commented on pkg/archive/archive_test.go:
```diff
@@ -4,6 +4,7 @@ import (
 	"archive/tar"
 	"bytes"
 	"fmt"
+	"github.com/docker/docker/pkg/testutil/assert"
```
nit but this import is not in the right place 😛

vdemeester reviewed (approved):
LGTM 🦁 
Just one small comment 👼
